### ISSUE TITLE
[Re] The Wisconsin Card Sorting Test: an implementation of Dehaene and Changeux Neuronal Network Model

### ISSUE THREAD

**Issue by PaulineBock** [ReScience/submissions#7]
**Original article:** Dehaene, S., & Changeux, J. P. (1991). The Wisconsin Card Sorting Test: Theoretical analysis and modeling in a neuronal network. Cerebral cortex, 1(1), 62-79.

**PDF URL:** https://github.com/PaulineBock/WCSTDehaeneChangeux/blob/master/ReScience_Dehaene.pdf
**Metadata URL:** https://github.com/PaulineBock/WCSTDehaeneChangeux/blob/master/Metadata.yaml
**Code URL:** https://github.com/PaulineBock/WCSTDehaeneChangeux

**Scientific domain:** Computational Neuroscience
**Programming language:** Python
**Suggested editor:**

**Comment by rougier** [comment]
@PaulineBock Thank your for your submission and sorry for the delay. We'll assign an editor soon.

**Comment by rougier** [comment]
@gdetor @benoit-girard @otizonaizit @oliviaguest Can any of you edit this submission? (I cannot, I've a conflict)

**Comment by oliviaguest** [comment]
Happy to do it.

**Comment by oliviaguest** [comment]
@PaulineBock there is a minor typo in the penultimate line of the introduction, I think it should be "implementations" (plural).

@rougier why is there no checkbox for the type of replication?

**Comment by PaulineBock** [comment]
Hello Olivia Guest,
thank you for the review, and yes I think it should be plural, I fix it
right now.



Le jeu. 5 sept. 2019 à 18:41, Olivia Guest <notifications@github.com> a
écrit :

**Comment by rougier** [comment]
@oliviaguest Thanks.

**Comment by oliviaguest** [comment]
@andreacaso do you think you can review this please?

**Comment by oliviaguest** [comment]
Also: @qihongl do you think you might be able to review this? 😊

**Comment by AndreaCaso** [comment]
Yes :) 

> @AndreaCaso do you think you can review this please? 😄

**Comment by AndreaCaso** [comment]
Hello Pauline, here's my review. 

**Code:** 

In order to run the model on PyCharm I had to make just one (very) minor change. I added an argument to the main function (WCST_test in the file DehaeneChangeux.py) to prevent the output window from showing up, so that whenever one runs 500 trials in a row (in Statistics.py) PyCharm won't freeze up. 
The code works very well, otherwise.

**Text:** 

'to evaluate frontal lobe dysfunction': This could be rephrased a little, because a number of publications argue that the WCST in not as sensitive or specific to frontal lobe lesions as once thought. 

'the subject is told to match the cards...': For a casual reader who is not familiar with the test, this is not very clear.  Perhaps citing Heaton (1981) could help direct the reader to the relevant literature

'processus': I believe this is a typo

'and a maximum value': what is this maximum value?

'without more precisions': could you please clarify?

'it is the reciprocal of speed of learning': why? could you please clarify? 

In addition:

- Figure 3 shows the comparison between the original paper and the replication. 
This is good, but the graph shows just one type of simulation (the implemented and modified network). What happened if one lesions the reward or the rule-coding network in the replicated model? 

Also: I believe that the reasoning + memory condition is equivalent to the auto-evaluation + memory, is that correct? It would be helpful to the reader if these concept where clarified and consistent nomenclature was used. 

- I understand that the modifications have been introduced in order to replicate the same results observed in the original paper because some details on the implementation weren't very clear (please tell me if I understand this correctly). If this is the case, I would also add a few lines in the Implementation and Modification paragraph and, most importantly, in the Conclusion, to describe the results obtained WITHOUT the modification and compare it against the one obtained WITH the modification. 
What can be gleaned by all this in terms of cognitive processes? 

I hope it helps. Feel free to ask if something is not clear.
Best, 
Andrea

**Comment by qihongl** [comment]
Wow, this is really interesting! Thanks for the replication. I'm happy to help! I will start reading it next week! 

I just looked at the code briefly. Just a quick minor comment: Making some object-oriented classes to separate the code for defining the model/task vs. the code for executing the experiment/plotting the results, might further enhance the readability and modularity of code :)

**Comment by PaulineBock** [comment]
Hello to all of you,

Thanks for the review. I will try to answer to all the questions.

I can rephrase the part about frontal lobe dysfunction and read more about
the clinical aims of the WCST.

I will cite Heaton and read more about it.

This is indeed a typo.

The maximum value depends of the long-term component values of the unit.

When I say that we have not more precisions, it is because on the original
paper, there is only one phrase about the activation of the go unit, where
the authors only says that it  is activated externally and that's all.
Maybe I could cite this phrase ?
"Propagation is gated by the activity of a "go" unit. Hence, a response is
given only when an external go signal is received. The go cluster might
also be activated by endogenous decision processes, but we do not use this
possibility in the following simulations."

For the reciprocal, I reused the term coming from the original paper, it is
to say that speed of learning and single-trial learning are closely related
and for which the aim is to show if the model can learn quickly by making
only one error.

I have already tried to test some lesioning before the code was finished,
and the model was not functioning, I am going to test again to lesion my
model to see what it does. (By lesioning reward and also lesioning memory
of rejected rules).

Yes, auto-evaluation and reasoning are the same thing I think.

I can talk about the first model without modifications, but in fact, there
were also some personal inventions, for the activation of the go unit, as
there was not any explanation in the paper about it.

And I agree to slit the code to see clearer.

This is the first time I am asking for a publication so sorry if I don't do
all things well, thank you for your review, I am going to make the changes,
I don't know if I should answer to your questions by mail or do directly
the changes on the git, tell me if I was right to answer you first.

Best,
Pauline.

**Comment by PaulineBock** [comment]
Sorry, I have just realised that for the "precisions" part, I used some
French word I think. I wanted to say "without more details"!

**Comment by oliviaguest** [comment]
@PaulineBock you are doing a great job and please do everything on github to keep the review process as open as possible. 😊

**Comment by qihongl** [comment]
Ah just to be clear I think the code is great! 

Actually I'm aware that refactoring the code might be a lot of work and I'm not sure if this is necessary for publishing this work. I was just imagining this work might be used in a cognitive modeling class in the future. I guess having another layer of abstraction can give students the option of forming some overall impression on how the code works first before diving into the detail.

**Comment by PaulineBock** [comment]
Hello to all of you,

I have made some modifications according to Andrea Caso's review. Tell me
if it is okay.
I have modified among other things the statistics graph to add the new
results obtained by lesioning the reward and the rule-coding clusters of
the implemented model.

Best regards,

Pauline.

Le lun. 9 sept. 2019 à 16:54, Andrea Caso <notifications@github.com> a
écrit :

**Comment by rougier** [comment]
Just step-in as editor-in-chief here: it seems the article includes image from the original article and to do that, we need permission from the original publisher (e.g. Elsevier, Springer, etc) or you'll need to redo the figure (and state "adapted from"). If it is already the case, don't pay attention to my message.

**Comment by PaulineBock** [comment]
Thanks for the warning Nicolas. So I have decided to redo the figure and
have stated "adapted from".
Best regards,
Pauline

**Comment by oliviaguest** [comment]
@AndreaCaso please see above! ☝️

**Comment by AndreaCaso** [comment]
Hello,

Thank you for making the changes. I have a few additional notes.

- Reference system should be consistent. The in-text citation for Heaton is (Heaton, R.K, (1981)), and the one for Dehaene and Changeux is [1]. If the reference system you have to abide to is the APA format I'd suggest that you use (Heaton, 1981) and (Dehaene & Changeux, 1991).

- The caption in Fig.3 could potentially be improved, so as to help the reader understand without necessarily referring to the text. I suggest something along these lines: "A comparison between the original model and our re-implementation: Each pattern in the grayscale bar graph represents a different model used  in the original paper (see legend). The bar graphs in colour represent the value obtained with our re-implementation. The overlap in the greyscale and colour bar graphs shows a direct comparison between the value obtained with our re-implementation and those obtained by the original model..."

A few other suggestions:
-  In Fig. 3: 'was succeed' is probably a typo.
- 'The test is no more succeed' could be rephrased.

Provided that the necessary changes are made, I recommend acceptance of the manuscript.

Regards,
Andrea

**Comment by qihongl** [comment]
Sorry about the delay :) will start reading the papers in the next few days!

**Comment by oliviaguest** [comment]
Thanks @qihongl! As much as you can please read it bearing in mind what else has been said above. 👍

**Comment by qihongl** [comment]
Thanks! @PaulineBock I really enjoyed reading it! Here're some thoughts. Please use them as you see fit. 

You used a lot of space to describe the order of computations in the model, which is good. Though I think the paper can focus more on conceptual points such as i) conceptually, what’s the original findings and how does their simulation results support that, and ii) what’re the most relevant aspects of the original results, or what's the most important patterns of capture in terms of a replication, and why your replication qualitatively captures that. 

I also think more discussion on why are those additional components needed to replicate the original results. In the absence of these additional mechanisms, how did the original implementation reach the same results? I know this is itself a difficult research question... Is it possible to contact the original author to get some advice/insights? 

The original paper is pretty old. Some terminologies might be unfamiliar to many folks. I think connecting them with terminologies can help many readers. I was wondering if other folks have the same feeling. For example, the phases "neuronal clusters/assemblies” are used in the paper many times. I think they are just neural network layers. In this case, pointing our that they are the same and stick with one word-choice might be better. 


#### minor comments/questions: 
"Firstly, all neuronal activities are set to 0 before network activation…”
Do you mean before a trial/or a stimuli block begins? 


“The criterion for a change of rule is 3 successes in a row” 
I think some might get confused about this — perhaps explicitly say something like "the rule changes whenever the agent made three consecutive correct assignment”?  

This is optional, but I would summarize the reward contingency as a tree diagram (instead of using the five bullet points). 

Since the model has many layers, the order of execution is not so easy to describe. I would put some numbers on the arrows (in the model figure) to make things very clear. 

For the new model figure, how about using a different color for the new components? 

I recommend choosing colors from the colorblind palette: https://seaborn.pydata.org/tutorial/color_palettes.html

**Comment by qihongl** [comment]
btw, the stats code generate plots as the program is being executed. I think it might better to save all plots to some image dir and suppress the output by default?

**Comment by PaulineBock** [comment]
Hello  @qihongl <https://github.com/qihongl>, thanks for the review.
Concerning the conceptual and discussion questions, I don't have enough
time to work on it, but I have transferred them to my co-author and
previous internship supervisor Frédéric Alexandre.
Concerning the other questions:
-yes we could tell that it is composed of several neural network layers
-the neuronal cluster activities are initially set to 0 before a whole test
of WCST and so before the trials, but not before each block stimuli.
-I will change the criterion explanation, yours is clearer.
-I have added the tree diagram, good idea.
- and the number of execution

- For which figure should I add/change the colors ?

Best,

Pauline.

Le lun. 30 sept. 2019 à 19:00, Qihong Lu <notifications@github.com> a
écrit :

**Comment by qihongl** [comment]
Thanks for the revision @PaulineBock I think it looks great. 

Right, I do think most readers would naturally ask why are there some differences between this implementation versus the original implementation. For example, is it possible that the original paper forgot to report something that's critical for that model to work? Is it possible that there the task implementation / stimuli representation used in the current implementation is slightly different from the original? etc. It would be ideal if there are some responses to this point in the paper. Again, I know this is a challenging point to address. I think contacting the original author to get some feedback might be useful. 

Besides that, I don't have additional major concerns. Thanks again!

**Comment by PaulineBock** [comment]
Hello @qihongl, so we have added some comment about the need of
modifications in the implementation according to the original model. It is
in the Modifiation and Implementation paragraph.

**Comment by oliviaguest** [comment]
What's left to be done here @PaulineBock? 😄

**Comment by PaulineBock** [comment]
@oliviaguest I think I have tried to answer to all your questions as I
could. Tell me if it seems okay for you !

Le lun. 7 oct. 2019 à 10:33, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
@qihongl @AndreaCaso can you please let me know what you think now with respect to what needs to be done here? For example, are you both proposing acceptance now?

**Comment by qihongl** [comment]
Sure -- If all of us @PaulineBock  @oliviaguest  @AndreaCaso  are happy with the final version. I think the revision is great. 

Just a really minor point: the blank space on page 6 is slightly awkward :) I was wondering if it can be reformatted slightly.

**Comment by AndreaCaso** [comment]
Hi, I think the revision is very good and I am happy with the final version.

**Comment by PaulineBock** [comment]
Indeed I have removed the space page 6 !
It is okay for me now.
Thank you all.

Best,
Pauline.

Le mar. 8 oct. 2019 à 10:09, Andrea Caso <notifications@github.com> a
écrit :

**Comment by rougier** [comment]
Any update ?

**Comment by oliviaguest** [comment]
Hey, I'm really sorry... new job and everything is so slow. I'll get on
this tomorrow. :)

Olivia

***
Dr Olivia Guest
sent from my smartphone

**Comment by oliviaguest** [comment]
@PaulineBock do you have an orcid? Please make one since you will be asked this so often it is worth just doing this now..?

**Comment by PaulineBock** [comment]
Hello, I have created my ORCID account. Here is my ID:
https://orcid.org/0000-0002-9374-8733

Thanks for your work.


Le jeu. 17 oct. 2019 à 08:47, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
@PaulineBock thank you ­— as soon as you merge https://github.com/PaulineBock/WCSTDehaeneChangeux/compare/master...oliviaguest:patch-1 I'll carry on :blush:

**Comment by PaulineBock** [comment]
I think it is okay ! Thank you really much!
Best,

Pauline.

Le jeu. 17 oct. 2019 à 10:23, Olivia Guest <notifications@github.com> a
écrit :

**Comment by qihongl** [comment]
Congrats! @PaulineBock 
Thanks folks! @oliviaguest @AndreaCaso

**Comment by rougier** [comment]
@oliviaguest Just a little help with the new publication process. @PaulineBock Can you complete your metadata file with the name/ORCID of the editor, reviewers, acceptance data and volume number (=5). As for the issue, I think we can stay on issue 1 or open a new issue (3) because there was a special issue (2). @oliviaguest what do you think?

**Comment by oliviaguest** [comment]
@rougier that sounds fine — thank you!

**Comment by PaulineBock** [comment]
Hello everyone, sorry for the late answer. So, do someone have all the
informations about the orcid of reviewers and editor ? I will edit as soon
as possible. Thanks.

Le lun. 21 oct. 2019 à 16:03, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
Mine is: 0000-0002-1891-0972

**Comment by PaulineBock** [comment]
So, I have found @AndreaCaso ORCID ID, but for @qihongl I have found 2
different orcid. What is the date of acceptance ?
Best.

Le lun. 4 nov. 2019 à 14:32, Olivia Guest <notifications@github.com> a
écrit :

**Comment by qihongl** [comment]
Sorry about the confusion :) 
here's my orcid: http://orcid.org/0000-0002-0730-5240

**Comment by PaulineBock** [comment]
Thank you, so metadata is completed for my part.
Tell me if there are some more things to do and when you will publish
please.

**Comment by oliviaguest** [comment]
@rougier what do I need, which directory do I need to be in, in order to run `process.py`? https://github.com/ReScience/articles/blob/master/README.md

**Comment by rougier** [comment]
Best is to clone the articles repository and to work from the root of that directory (with the PDF and metadata). You can finish filling the metadata and do not forget to use the Zenodo sandbox first to check everythign's fine.

By the way, don't try to publish the website once you've added the new entry into published, just drop me an email or a notification.

**Comment by oliviaguest** [comment]
@PaulineBock @rougier 
I think this is what is needed to be updated in the metadata, correct? 


```
Article ID:  416801
Article DOI: 10.5072/zenodo.416801
Article URL: https://sandbox.zenodo.org/record/416801/files/ReScience_Dehaene.pdf
```

**Comment by rougier** [comment]
Yes the sandbox server is only for you to check everything ran ok.  Since this seems to be the case, you can no run it on the real server and communicate the information to @PaulineBock. Note that the alternative is to modify the metadata yourself and upload the (recompiled) paper and to make a PR to @PaulineBock repo.

Once finished, don't try to "publish" the rescience website since it may fail if you do not have all the plugins. Just add the new bibtex entry to the _published.bib and drop me a notification or an email.

**Comment by oliviaguest** [comment]
I'll give this a shot on Monday — thank you @rougier!

**Comment by oliviaguest** [comment]
@rougier @PaulineBock I think/hope I have done this right:

**Comment by oliviaguest** [comment]
[Sorry, did not meet to close this issue. I keep switching from Mac to Linux which means the shortcuts are confused in my head.]

**Comment by rougier** [comment]
Looks good to me. But it's not yet online, right ?

**Comment by oliviaguest** [comment]
I don't think so, the URL 404s and I didn't do anything other than run what is asked for in step 1.

**Comment by oliviaguest** [comment]
@rougier what's the next step? does @PaulineBock have to do something or is it me? Sorry, I'm not 100% clear on the next step.

**Comment by rougier** [comment]
If you've made the pre-registration on the real Zenodo server, then you or @PaulineBock have to update the metadata file (with DOI, editor name, reviewer name, etc) and regenerate the PDF accordingly (the DOI should now appear on the PDF). Once done, you can register the new PDF at Zenodo still using the script. In case of problem, you still have the option  to manually change the Zenodo entry through the Zenodo website.

Tell me if something's note clear and feel free to improve the description of the process. I realize that I test it for myself knowing the process but in the end the instructions might be a bit confusing.

**Comment by oliviaguest** [comment]
Yes, I did it on the real server. I'll have a look at the instructions and send a PR if needed. Thank you!

**Comment by oliviaguest** [comment]
Does this look right now, @rougier? https://github.com/PaulineBock/WCSTDehaeneChangeux/pull/2

**Comment by rougier** [comment]
LGTM. Do not forget other informatipon (journal, issue, number, editor, reviewer, etc)

**Comment by oliviaguest** [comment]
@rougier OK, as soon as the PR is merged. Where do I find the ISSN?

@PaulineBock can you fill in the bits that are for you to do, please?|


```
# Code URL and DOI (url is mandatory for replication, doi after acceptance)
# You can get a DOI for your code from Zenodo,
#   see https://guides.github.com/activities/citable-code/
code:
  - url: https://github.com/PaulineBock/WCSTDehaeneChangeux
  - doi:
```

and:

```
# To be filled by the author(s) after acceptance
# -----------------------------------------------------------------------------

# For example, the URL of the GitHub issue where review actually occured
review: 
  - url:
```

**Comment by PaulineBock** [comment]
@oliviaguest  hello ! I've just updated the metadata, tell me if it's okay.

**Comment by oliviaguest** [comment]
Great, @PaulineBock — can you merge the PR too, please? https://github.com/PaulineBock/WCSTDehaeneChangeux/pull/2

**Comment by PaulineBock** [comment]
Done.

Le mar. 26 nov. 2019 à 11:08, Olivia Guest <notifications@github.com> a
écrit :

**Comment by rougier** [comment]
ISSN 2430-3658 (bottom of ReScience website)

**Comment by oliviaguest** [comment]
All this info should be on one easy access place, perhaps? 

@PaulineBock I think this is the final edit? https://github.com/PaulineBock/WCSTDehaeneChangeux/pull/3

**Comment by PaulineBock** [comment]
I think yes. Thanks for all your work.

Best.

Le jeu. 28 nov. 2019 à 13:43, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
@PaulineBock sorry! this is also needed! https://github.com/PaulineBock/WCSTDehaeneChangeux/pull/4

**Comment by PaulineBock** [comment]
Is it good now? I have merged the 3 last modifications. (I have marged the
first before seeing that it was a mistake)

Le jeu. 28 nov. 2019 à 14:09, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
@PaulineBock yes, it looks good — thank you! :smile_cat: 

@rougier is everything OK to actually publish to zenodo now?

**Comment by rougier** [comment]
No, the type of the article is not editorial but replication and the "under review" should not appear anymore in the PDF. There are some comments in the original metadata yaml file to help filling it.

**Comment by oliviaguest** [comment]
I'm really confused. Type of article is replication? What's wrong with that? https://github.com/PaulineBock/WCSTDehaeneChangeux/blob/11fd0458ceb000329018733d40a59aa9d483e186/Metadata.yaml#L106

**Comment by PaulineBock** [comment]
For the pdf I must fill Metadata.tex and push pdf again as I work with
overleaf. I am completing the changes.

Le jeu. 28 nov. 2019 à 16:40, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
The type seems correct in the YAML.

**Comment by PaulineBock** [comment]
Yes but I think I had seen that when working on overleaf  the Metadata.tex
cannot be automatically generated from the Metadata.yaml
Not sure but in case I have edited the metadata.tex myself according to the
metadata.yaml.

So now it should be good.

Le jeu. 28 nov. 2019 à 17:10, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
Not sure what you need to do over what I need to do. :sweat_smile: I think it's all you now @PaulineBock?

**Comment by rougier** [comment]
@PaulineBock There's a python script to convert the metadata.yaml into metadata.tex but else, you can also do it manually.

**Comment by PaulineBock** [comment]
I have modified it manually. Is it okay for the pdf?

**Comment by rougier** [comment]
Not quite yet. I think you're missing the domain (Computational Neuroscience) and you'll need to replace subsection by section (such as to have 1 Introduction instead of 0.1 Introduction)

**Comment by oliviaguest** [comment]
@PaulineBock how can I compile your PDF?

**Comment by oliviaguest** [comment]
Edits that need to made (based on @rougier above and me reading through the PDF now):

- [ ] ORCIDs need to be fixed as they currently all point to mine.
- [ ] Domain needs to be Computational Neuroscience.
- [ ] Replace subsection by section (such as to have 1 Introduction instead of 0.1 Introduction).
- [ ] Published date needs to be amended since it didn't get published yet. Would be easy to fix if I could compile your PDF myself so that I can add that in when I publish it.

**Comment by PaulineBock** [comment]
As I said before I was using overleaf, I can share you my project if you
want. How can I send you a private message ?

Le lun. 2 déc. 2019 à 11:53, Olivia Guest <notifications@github.com> a
écrit :

**Comment by PaulineBock** [comment]
I have just edited all that was necessary and pushed on github. If we
publish today it is good, else I will give you the link to edit the
overleaf in private to change the date of publication.

**Comment by oliviaguest** [comment]
Email me the link if you can, please? :smile:

**Comment by PaulineBock** [comment]
I don't have your email and is this issue accessible to public ? Or anyone else can use the link reading this issue?

**Comment by oliviaguest** [comment]
contact@oliviaguest.com

**Comment by oliviaguest** [comment]
@PaulineBock could you also give me push access to your repo https://github.com/PaulineBock/WCSTDehaeneChangeux/ so I can upload the final PDF, please? :blush:

**Comment by oliviaguest** [comment]
This is how: https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/inviting-collaborators-to-a-personal-repository

**Comment by PaulineBock** [comment]
Thanks for the tips, it is done :)

Le mer. 4 déc. 2019 à 13:04, Olivia Guest <notifications@github.com> a
écrit :

**Comment by oliviaguest** [comment]
Sorry for the delay, just working through some stuff @PaulineBock! 

@rougier when you get a chance please feel free to advise... :sweat_smile:  https://github.com/ReScience/articles/issues/2

**Comment by rougier** [comment]
Just did.

**Comment by oliviaguest** [comment]
Woo hoo! https://zenodo.org/record/3545905

Finally! :smile:

**Comment by oliviaguest** [comment]
Congratulations @PaulineBock! :tada:

**Comment by PaulineBock** [comment]
Thank to all of you!! It was a pleasure to work with you :)

Le lun. 9 déc. 2019 à 15:41, Olivia Guest <notifications@github.com> a
écrit :
